allennlp/data/dataset_readers/universal_dependencies_multilang.py

```diff
@@ -0,0 +1,171 @@
+
+    Parameters
+    ----------
+    token_indexers : ``Dict[str, TokenIndexer]``, optional (default=``{"tokens": SingleIdTokenIndexer()}``)
+        The token indexers to be applied to the words TextField.
+    use_language_specific_pos : ``bool``, optional (default = False)
+        Whether to use UD POS tags, or to use the language specific POS tags
+        provided in the conllu format.
+    alternate : ``bool``, optional (default = True)
+        Whether to alternate between input files.
+    is_first_pass_for_vocab : ``bool``, optional (default = True)
+        Whether the first pass will be for generating the vocab. If true,
+        the first pass will run over the entire dataset of each file (even if alternate is on).
+    batch_size : ``int``, optional (default = 32)
+        The amount of consecutive cases to sample from each input file when alternating.
+    """
+    def __init__(self,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 use_language_specific_pos: bool = False,
+                 lazy: bool = False,
+                 alternate: bool = True,
+                 is_first_pass_for_vocab: bool = True,
+                 batch_size: int = 32) -> None:
+        super().__init__(lazy)
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+        self.use_language_specific_pos = use_language_specific_pos
+
+        self.is_first_pass_for_vocab = is_first_pass_for_vocab
+        self.alternate = alternate
+        self.batch_size = batch_size
+
+        self.is_first_pass = True
+        self.iterators = None
+        self.instances_per_lang = defaultdict(int)
+
+    def _read_one_file(self, lang: str, file_path: str):
+        with open(file_path, 'r') as conllu_file:
+            logger.info("Reading UD instances for %s language from conllu dataset at: %s", lang, file_path)
+
+            for annotation in  lazy_parse(conllu_file.read()):
+                # CoNLLU annotations sometimes add back in words that have been elided
+                # in the original sentence; we remove these, as we're just predicting
+                # dependencies for the original sentence.
+                # We filter by None here as elided words have a non-integer word id,
+                # and are replaced with None by the conllu python library.
+                annotation = [x for x in annotation if x["id"] is not None]
+
+                heads = [x["head"] for x in annotation]
+                tags = [x["deprel"] for x in annotation]
+                words = [x["form"] for x in annotation]
+                if self.use_language_specific_pos:
+                    pos_tags = [x["xpostag"] for x in annotation]
+                else:
+                    pos_tags = [x["upostag"] for x in annotation]
+                yield self.text_to_instance(lang, words, pos_tags, list(zip(tags, heads)))
+
+    @overrides
+    def _read(self, file_paths: Dict[str, str]):
```

Comment:
I see you just did a fix for this, and your choice looks good to me.  Yesterday I found another example of how I handled this: https://github.com/matt-gardner/allennlp/blob/77b066516b8b12de8521dc8a431741555937917f/allennlp/data/dataset_readers/visual_question_answering/shapes_reader.py#L21-L26.  Not saying you should change what you did, just another example in case you find it useful.  I think that what you did is probably better, though, as my version still has problems at test time.